src/python/pants/engine/internals/parametrize.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import itertools
+from dataclasses import dataclass
+from typing import Any, Iterator
+
+from pants.engine.addresses import Address
+from pants.util.meta import frozen_after_init
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class Parametrize:
+    """A builtin function/dataclass that can be used to parametrize Targets.
+
+    Parameterization is applied between TargetAdaptor construction and Target instantiation, which
+    means that individual Field instances need not be aware of it.
+    """
+
+    args: tuple[str, ...]
+    kwargs: dict[str, Any]
+
+    def __init__(self, *args: str, **kwargs: Any) -> None:
+        self.args = args
+        self.kwargs = kwargs
+
+    def to_parameters(self) -> dict[str, Any]:
+        """Validates and returns a mapping from aliases to parameter values.
+
+        This conversion is executed lazily to allow for more context in error messages, such as the
+        TargetAdaptor consuming the Parametrize instance.
+        """
+        parameters = dict(self.kwargs)
+        for arg in self.args:
+            if not isinstance(arg, str):
+                raise Exception(
+                    f"In {self}:\n  Positional arguments must be strings, but "
+                    f"`{arg}` was a `{type(arg).__name__}`."
+                )
+            previous_arg = parameters.get(arg)
+            if previous_arg is not None:
+                raise Exception(
+                    f"In {self}:\n  Positional arguments cannot have the same name/alias as "
+                    f"keyword arguments. `{arg}` was also defined as `{arg}={previous_arg}`."
+                )
+            parameters[arg] = arg
+        return parameters
```

Comment:
Ditto expectation, to return instance of `FrozenDict`.

Reply:
I'll make this private, since the expectation is only that it is called as part of `expand`.

Reply:
Ended up leaving this public for testing purposes.